### PR TITLE
Unpin host memory via host pointer instead of GPU pointer

### PR DESCRIPTION
**Pull request by whchung**
This commit fixes to logic in HCC runtime to be compatible with the semantic
asked by ROCR runtime API: hsa_amd_memory_unlock.

### CODE REVIEW

**Review by scchan — APPROVED** [review]
LGTM

**Comment by fxkamd** [comment]
Unpinning with the GPU pointer should work. There was a bug in the Thunk that is being fixed right now.

Using the CPU pointer works OK as long as you never pin the same address multiple times with different sizes. Otherwise you'll get different GPU pointers for the different mappings, and the only way to unpin the correct one is to use the GPU pointer.

**Comment by AlexVlx** [comment]
@fxkamd OK, I'm confused now. The HSA extension itself is specified in terms of the host accessible address, see https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/bfc4b9e98cb5e48d9f96287371d518cb444968ba/src/inc/hsa_ext_amd.h#L1205. Are we going to change that? Furthermore, do we not pin at page granularity, which basically entails that you'd only get different GPU pointers when crossing pages? @hthangirala, can you comment here?

P.S.: strictly in terms of taste and ease of understanding, unpinning the GPU pointer seems significantly more natural, and lends itself to easier wrapping in e.g. unique_ptr, so it would be nice if that were guaranteed to work, however the current specification does not reflect that intent, Thunk bug notwithstanding.

**Comment by fxkamd** [comment]
Pinning and unpinning of host memory supports unaligned pointers and sizes. They don't need to be page granularity.

We had to make some subtle changes to the semantics of the APIs to support OpenCL use cases allowing the same pointer to be pinned multiple times with proper reference counting etc. In some cases the same pointer pinned multiple times can return different GPU pointers. Therefore the CPU pointer cannot be used reliably as a unique identifier for unpinning. This may not have been reflected properly in the spec.

If you allow pinning the same pointer multiple times, potentially with different sizes, I recommend you use the GPU pointer for unpinning.

**Comment by AlexVlx** [comment]
@fxkamd, I wasn't thinking about the user's perspective here, but rather the actual side effect of pinning. Are we doing something beyond what the kernel does here because mlock works at page granularity last unless memory betrays me; this is honest curiosity as from a runtime implementer's perspective these are relevant nuances.

**Comment by fxkamd** [comment]
Right, pinning and mapping memory for GPU access works at page granularity. But the Thunk will do the necessary rounding and return a GPU pointer with the same offset from the start of the page. So these APIs can be used with arbitrary pointers.

**Comment by AlexVlx** [comment]
Thank you for the clarification, much appreciated.

**Comment by scchan** [comment]
@AlexVlx this needs your review and approval, could you take a look?

**Comment by AlexVlx** [comment]
@scchan @whchung My understanding is that this is getting fixed in Thunk (per @fxkamd) so that the current code would be correct (and potentially more flexible). I've not yet reasoned fully whether or not one can construct a scenario where a global is aliased by another global of a different size (seems doable if unlikely). I'm not opposed to this change per se, but we probably need to keep the above on the radar, and possibly revert in a version or two.

**Review by AlexVlx — APPROVED** [review]
LGTM modulo the comment I provided elsewhere.

**Pull request merged**
